CWE-016 - Adds MaxRequestLength check for ASP.NET

cldrn opened the pull request:
![maxrequestlength](https://user-images.githubusercontent.com/191724/67532526-24841100-f68c-11e9-8a03-6715e3429ccc.PNG)

hvitved commented on csharp/ql/src/Security Features/CWE-016/ASPNetMaxRequestLength.qhelp:
```diff
@@ -0,0 +1,22 @@
+
+<overview>
+<p>
+The MaxRequestLength attribute sets the limit for the input stream buffering threshold in KB. Attackers can use large files to cause denial of service attacks.
```
`MaxRequestLength` -> `<code>MaxRequestLength</code>`

hvitved commented on csharp/ql/src/Security Features/CWE-016/ASPNetMaxRequestLength.ql:
```diff
@@ -0,0 +1,18 @@
+/**
+ * @name Large maxRequestLength value.
```
No period at the end.

hvitved commented on csharp/ql/src/Security Features/CWE-016/ASPNetMaxRequestLength.ql:
```diff
@@ -0,0 +1,18 @@
+ * @kind problem
+ * @problem.severity warning
+ * @precision very-high
+ * @id cs/web/debug-binary
```
`cs/web/debug-binary` -> `cs/web/max-request-length`

hvitved commented on csharp/ql/src/Security Features/CWE-016/ASPNetMaxRequestLength.ql:
```diff
@@ -0,0 +1,18 @@
+import csharp
+from XMLAttribute a
+where
+  a.getName().toLowerCase() = "maxrequestlength" and a.getValue() > "4096"
```
The test `a.getValue() > "4096"` actually performs a lexicographic string comparison; `a.getValue().toInt() > 4096` is better.

hvitved commented on csharp/ql/src/Security Features/CWE-016/ASPNetMaxRequestLength.ql:
```diff
@@ -0,0 +1,18 @@
+from XMLAttribute a
+where
+  a.getName().toLowerCase() = "maxrequestlength" and a.getValue() > "4096"
+select a, "maxRequestLength is set too large: "+a.getValue()+" KB. It is recommended to set it as minimum as possible based on business requirements."
```
`select a, "Large 'maxRequestLength' value (" + a.getValue() + ")."`

hvitved commented on csharp/ql/src/Security Features/CWE-016/ASPNetMaxRequestLength.ql:
```diff
@@ -0,0 +1,18 @@
+ * @description maxRequestLength should be set as minimum as possible.
+ * @kind problem
+ * @problem.severity warning
+ * @precision very-high
```
Please remove the precision tag (we will make sure to readd it in a follow-up PR.)

cldrn commented:
Thanks for the feedback. I've address all your comments. I wrote them while analyzing my first project so I'm still reading up on the documentation. 

Awesome work!

jf205 commented on csharp/ql/src/Security Features/CWE-016/ASPNetMaxRequestLength.ql:
```diff
@@ -0,0 +1,23 @@
+/**
+ * @name Large 'maxRequestLength' value
+ * @description Setting a large 'maxRequestLength' value may render a web page vulnerable to
```
```suggestion
 * @description Setting a large 'maxRequestLength' value may render a webpage vulnerable to
```

jf205 commented on csharp/ql/src/Security Features/CWE-016/ASPNetMaxRequestLength.qhelp:
```diff
@@ -0,0 +1,36 @@
+<sample src="Web.config.bad" />
+
+<p>
+Unless such a high value is strictly needed, it is better to set a lower value, for example 4096:
```
Suggest using consistent units (KB) in the examples.

jf205 commented on csharp/ql/src/Security Features/CWE-016/ASPNetMaxRequestLength.qhelp:
```diff
@@ -0,0 +1,36 @@
+<recommendation>
+
+<p>
+The recommended value is 4096 but you should try setting it as small as possible according to business requirements.
```
Perhaps change this sentence to:

The recommended value is 4096 _KB_, but you should set it as small as possible according to business requirements.

(Note, include units)

jf205 commented on change-notes/1.23/analysis-csharp.md:
```diff
@@ -9,6 +9,7 @@ The following changes in version 1.23 affect C# analysis in all applications.
 | **Query**                   | **Tags**  | **Purpose**                                                        |
 |-----------------------------|-----------|--------------------------------------------------------------------|
 | Unsafe year argument for 'DateTime' constructor (`cs/unsafe-year-construction`) | reliability, date-time | Finds incorrect manipulation of `DateTime` values, which could lead to invalid dates. |
+| Large 'maxRequestLength' value (`cs/web/large-max-request-length`) | security, frameworks/asp.net, external/cwe/cwe-16 | Finds `web.config` files with large (greater than 4096) `maxRequestLength` attributes. |
```
```suggestion
| Large 'maxRequestLength' value (`cs/web/large-max-request-length`) | security, frameworks/asp.net, external/cwe/cwe-16 | Finds `web.config` files with large (greater than 4096 KB) `maxRequestLength` attributes. |
```

jf205 commented on csharp/ql/src/Security Features/CWE-016/ASPNetMaxRequestLength.qhelp:
```diff
@@ -0,0 +1,36 @@
+<sample src="Web.config.good" />
+
+</example>
+
```
Is there a reference to some authoritative docs you could include here?

cldrn commented:
Closed this issue as I removed that local copy and started doing each PR in their own branch. At this point I have deleted the original branch for that PR. I've addressed the comments in there but can we keep working on this one https://github.com/Semmle/ql/pull/2355?

The other fork had a  PR on master branch which became a mess when I added others there by mistake. Hence, why I started from stratch with each PR in their own branch.

:)

hvitved commented:
Closing in favor of https://github.com/Semmle/ql/pull/2355.
